feat: add WASM compatibility for verification functions

sebasti810 opened the pull request:
This pull request introduces WebAssembly (WASM) compatibility for the verification process of Jolt. The main changes include:

- Extension of `jolt new` command with `--wasm` flag:
    - The `jolt new <project-name> --wasm` command modifies both the `root/Cargo.toml` and `guest/Cargo.toml` files to include the necessary dependencies for WASM. It also marks the project as a `cdylib` crate type, which is required for WASM compilation.
- Addition of `jolt build-wasm` command
   - The `build-wasm` command extracts all functions declared with `jolt::provable`. It then performs preprocessing steps to build and decode the program, storing the resulting data in binary form. An example `index.html` file is created to demonstrate how to use the functions generated by the macro extension. Finally, optimized WASM files are created using `wasm-pack` to make the project ready for use. 
 - Extension of macros:
    - The macros are enhanced to check if functions are marked with `jolt::provable(wasm)`. If so, necessary imports and helper functions are defined. The output of the preprocessing process is then used to build the WASM-compatible verify-functions. 

**Current Limitations**
- Dependencies used in `guest/src/lib.rs` need to be defined both in `guest/Cargo.toml` and `root/Cargo.toml`. This redundancy is known and could be optimized in future iterations.
- Currently, this only provides WASM support for proof verification. The proving itself is not yet WASM-compatible, bc it's unclear whether this is a desired requirement.

#314

ncitron commented on jolt-sdk/macros/src/lib.rs:
```diff
@@ -11,19 +11,42 @@ use common::{
 };
 
+lazy_static! {
+    static ref WASM_IMPORTS_ADDED: Mutex<bool> = Mutex::new(false);
```
Why are we using a lazy static here? We have many other instances where we dynamically generate code based on an attribute in the provable macro and we don't need to use these.

sebasti810 replied:
I thought we could use lazy_static here because the macro code for WASM needs certain global imports and a helper structure for serialization to be initialized only once, regardless of how many functions are labelled with the `wasm` attribute..

Do you think the global initialization is a problem? If lazy_static is the issue, we could use Once from the std-library like this:
```
static WASM_IMPORTS_INIT: Once = Once::new();

#[proc_macro_attribute]
pub fn provable(attr: TokenStream, item: TokenStream) -> TokenStream {
    let attr = parse_macro_input!(attr as AttributeArgs);
    let func = parse_macro_input!(item as ItemFn);
    let mut builder = MacroBuilder::new(attr, func);

    let mut token_stream = builder.build();

    // Add wasm utilities and functions if the function is marked as wasm
    if builder.has_wasm_attr() {
        // wasm utilities should only be added once
        WASM_IMPORTS_INIT.call_once(|| {
            let wasm_utilities: TokenStream = builder.make_wasm_utilities().into();
            token_stream.extend(wasm_utilities);
        });
        let wasm_token_stream: TokenStream = builder.make_wasm_function().into();
        token_stream.extend(wasm_token_stream);
    }

    token_stream
}
```

Would this approach be more suitable for our use case?

ncitron replied:
Oh I see what you're doing here now. I think this is fine.

ncitron commented on src/main.rs:
```diff
@@ -146,6 +160,240 @@ fn display_sysinfo() {
+fn preprocess_and_save(func_name: &str) -> Result<()> {
+    let mut program = Program::new("guest");
+    program.set_func(func_name);
+    program.set_std(false);
```
Can we figure out how to extract these options from the macro? This is how we currently do it for normal usage from the macro generated functions, which I am sure we could extract if needed.

sebasti810 replied:
How would it be if we also read out whether the feature is set or not in the preprocessing, in which we modify Cargo.toml anyway?

sebasti810 replied:
Also, for the todo comment below I thought about extracting the parse_attributes function,  for example into "common". This way we could reuse it both in preprocessing and in the macro. What do you think of this idea?

ncitron replied:
Yeah I think that is fine. It's just important that when we use this preprocess function it behaves the same way as it would in the non-wasm variants so that the prover/verifier are interoperable. Otherwise a proof generated would set the various attributes here but the wasm verifier would have these defaults instead.

sebasti810 commented:
@ncitron I have refactored the parse_attributes function so that we can use it in both the WASM preprocessing and within the macro. Additionally, I have moved the entire WASM preparation process to a separate file because it was getting very cluttered. Whether it uses std or no_std is determined by reading the features and setting it accordingly. lmk what you think

ncitron reviewed (approved):
LGTM!

ncitron commented:
Looks like CI is failing due to clippy. Can you resolve then we can merge?

sebasti810 commented:
@ncitron should be done
